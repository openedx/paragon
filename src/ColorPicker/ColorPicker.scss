@import "variables";

.pgn__color-picker {
  box-shadow: 0 0 3px 1px $gray-300;
  border: none;
  background:
    linear-gradient(to bottom right,
    rgb(255, 0, 0) 0%,
    rgb(255, 154, 0) 17%,
    rgb(208, 222, 33) 33%,
    rgb(79, 220, 74) 40%,
    rgb(28, 127, 238) 75%,
    rgb(95, 21, 242) 86%,
    rgb(186, 12, 248) 100%
  );

  &.pgn__color-picker-sm {
    height: $picker-size-sm;
    width: $picker-size-sm;
  }

  &.pgn__color-picker-md {
    height: $picker-size-md;
    width: $picker-size-md;
  }
}

.pgn__color-modal {
  background: $white;
  padding-bottom: .1rem;
  margin-bottom: 1rem;
}

.pgn__color-error {
  font-size: 70%;
}

.pgn__hex-form {
  display: flex;
  max-width: 180px;
  margin: 5px;

  h5 {
    margin: 10px;
  }

  .form-field {
    @media (min-width: map-get($grid-breakpoints, "sm")) {
      margin-inline-end: 0;
    }
  }
}
