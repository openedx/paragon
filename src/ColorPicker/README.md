---
title: 'ColorPicker'
type: 'component'
components:
- ColorPicker
categories:
- Imagery & Iconography
status: 'In progress'
designStatus: 'In progress'
devStatus: 'In progress'
notes: |
---

### Basic Usage

```jsx live
() => {
  const [colorSm, setColorSm] = useState('')
  const [colorMd, setColorMd] = useState('')

  return (
  <div className="d-flex bg-light p-2" style={{ columnGap: '1rem' }}>
    <ColorPicker color={colorSm} setColor={setColorSm} size="sm" />
    <ColorPicker color={colorMd} setColor={setColorMd} size="md" />
  </div>
  )
}
```
### With good and malformatted default color passed

```jsx live

() => {
  const [color, setColor] = useState('#32116c')
  const [colorBad, setColorBad] = useState('#badcolor')

  return (
  <div className="d-flex bg-light p-2" style={{ columnGap: '1rem' }}>
    <ColorPicker color={color} setColor={setColor} />
    <ColorPicker color={colorBad} setColor={setColorBad} />
  </div>
  )
}
```

