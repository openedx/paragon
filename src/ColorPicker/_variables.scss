$picker-size-sm:                     2rem !default;
$picker-size-md:                     calc(1.3333em + 1.125rem + 2px) !default;
