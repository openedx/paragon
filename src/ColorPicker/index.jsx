import React, { useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { HexColorPicker } from 'react-colorful';

import Button from '../Button';
import Form from '../Form';
import ModalPopup from '../Modal/ModalPopup';
import { OverlayTrigger } from '../Overlay';
import Tooltip from '../Tooltip';
import useToggle from '../hooks/useToggle';

function ColorPicker({
  color, setColor, className, size,
}) {
  const [isOpen, open, close] = useToggle(false);
  const [target, setTarget] = React.useState(null);
  const [hexValid, setHexValid] = React.useState(true);

  const validateHex = useCallback((input) => {
    const hexRegex = /^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/;
    if (input.length > 1 && !input.startsWith('#')) {
      setColor(`#${input}`);
    } else {
      setColor(input);
    }
    if (input === '' || hexRegex.test(input) === true) {
      setHexValid(true);
    } else {
      setHexValid(false);
    }
  }, [setColor]);

  // this is needed for when a user changes the color through the sliders
  useEffect(() => validateHex(color), [validateHex, color]);

  return (
    <>
      <span className="d-flex">
        <OverlayTrigger
          placement="top"
          overlay={<Tooltip id="color-picker-tooltip">Color picker</Tooltip>}
        >
          <Button
            ref={setTarget}
            className={classNames(
              className,
              'pgn__color-picker',
              `pgn__color-picker-${size}`,
            )}
            style={{
              ...(color && hexValid ? { background: `${color}` } : {}),
            }}
            onClick={open}
          />
        </OverlayTrigger>
      </span>
      <ModalPopup
        positionRef={target}
        isOpen={isOpen}
        style={{ background: 'black' }}
        onClose={close}
      >
        <div
          className="pgn__color-modal rounded shadow"
          style={{ textAlign: 'start' }}
        >
          <HexColorPicker color={color || ''} onChange={setColor} />
          <Form.Group className="pgn__hex-form" size="sm">
            <h5>Hex</h5>
            <Form.Control
              className="form-field"
              isInvalid={!hexValid}
              value={color}
              onChange={(e) => validateHex(e.target.value)}
              data-testid="hex-input"
            />
            {!hexValid && (
              <Form.Control.Feedback
                className="pgn__color-error"
                type="invalid"
              >
                Colors must be in hexadecimal format.
              </Form.Control.Feedback>
            )}
          </Form.Group>
        </div>
      </ModalPopup>
    </>
  );
}

ColorPicker.defaultProps = {
  color: '',
  className: undefined,
  size: 'md',
};

ColorPicker.propTypes = {
  /** A default hex code to preset the picker to display. */
  color: PropTypes.string,
  /** Passing setState function allows parent to alter the color. */
  setColor: PropTypes.func.isRequired,
  /** A class name to append to the base element. */
  className: PropTypes.string,
  /** Size of the color picker */
  size: PropTypes.oneOf(['sm', 'md']),
};

export default ColorPicker;
