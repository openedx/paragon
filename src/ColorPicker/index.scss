@import "variables";

.pgn__color-picker {
  box-shadow: 0 0 3px 1px $gray-300;
  border: none;
  background:
    linear-gradient(to bottom right,
    rgb(255, 0, 0) 0%,
    rgb(255, 154, 0) 17%,
    rgb(208, 222, 33) 33%,
    rgb(79, 220, 74) 40%,
    rgb(28, 127, 238) 75%,
    rgb(95, 21, 242) 86%,
    rgb(186, 12, 248) 100%
  );

  &.pgn__color-picker-sm {
    height: $picker-size-sm;
    width: $picker-size-sm;
  }

  &.pgn__color-picker-md {
    height: $picker-size-md;
    width: $picker-size-md;
  }
}

.pgn__color-modal {
  background: $white;
  padding-bottom: .1rem;
  margin-bottom: 1rem;
}

.pgn__color-error {
  font-size: 70%;
  flex: 1 1 auto;
  margin-left: .5rem;
}

.pgn__hex-label {
  font-weight: bold;
  font-size: $h5-mobile-font-size;
  padding: .5rem;
  margin-bottom: 0 !important;
}

.pgn__hex-form {
  display: flex;
  max-width: 180px;
  margin: 5px;
  flex-wrap: wrap;

  div {
    display: inline-flex;
  }

  .pgn__hex-field {
    @media (min-width: map-get($grid-breakpoints, "sm")) {
      margin-inline-end: 0;
    }
  }
}
