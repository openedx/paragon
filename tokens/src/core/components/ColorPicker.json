{
  "size": {
    "$type": "dimension",
    "color-picker": {
      "sm": {
        "$value": "2rem",
        "source": "$picker-size-sm"
      },
      "md": {
        "$value": "calc(1.3333em + 1.125rem + 2px)",
        "source": "$picker-size-md"
      }
    }
  }
}
