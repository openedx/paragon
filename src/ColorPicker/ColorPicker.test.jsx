import React from 'react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { render, fireEvent, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import '@testing-library/jest-dom';
import ColorPicker from './index';

describe('renders correctly', () => {
  const color = '';
  const setColor = jest.fn();
  it('successfully renders', () => {
    const tree = renderer.create(<ColorPicker color={color} setColor={setColor} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders with defaults', () => {
    const { container } = render(<ColorPicker color={color} setColor={setColor} />);
    expect(container.firstChild.firstChild).toHaveClass('pgn__color-picker-md');
  });
  it('renders with the passed variables ', () => {
    const className = 'testClassName';
    const { container } = render(
      <ColorPicker color={color} setColor={setColor} className={className} size="sm" />,
    );
    expect(container.firstChild.firstChild).toHaveClass(className);
    expect(container.firstChild.firstChild).toHaveClass('pgn__color-picker-sm');
  });
});
describe('picker works as expected', () => {
  const color = 'wassap';
  const setColor = jest.fn();
  it('validates hex color', async () => {
    render(<ColorPicker color={color} setColor={setColor} />);
    await act(async () => {
      userEvent.click(screen.getByRole('button'));
    });
    expect(screen.queryByText('Colors must be in hexadecimal format.')).toBeInTheDocument();
    const input = screen.getByTestId('hex-input');
    fireEvent.change(input, { target: { value: '#32116c' } });
    expect(screen.queryByText('Colors must be in hexadecimal format.')).not.toBeInTheDocument();
  });
});
