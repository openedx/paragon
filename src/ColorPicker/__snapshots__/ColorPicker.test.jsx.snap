// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly successfully renders 1`] = `
<span
  className="d-flex"
>
  <button
    className="pgn__color-picker pgn__color-picker-md btn btn-primary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    style={Object {}}
    type="button"
  />
</span>
`;
